src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -229,17 +271,96 @@ public Callable<?> getInitializationTask()
     {
         // New storage-attached indexes will be available for queries after on disk index data are built.
         // Memtable data will be indexed via flushing triggered by schema change
-        return () -> startInitialBuild(baseCfs).get();
+        // We only want to validate the index files if we are starting up
+        return () -> startInitialBuild(baseCfs, StorageService.instance.isStarting()).get();
     }
 
-    private Future<?> startInitialBuild(ColumnFamilyStore baseCfs)
+    private Future<?> startInitialBuild(ColumnFamilyStore baseCfs, boolean validate)
     {
         if (baseCfs.indexManager.isIndexQueryable(this))
         {
             logger.debug(indexContext.logMessage("Skipping validation and building in initialization task, as pre-join has already made the storage attached index queryable..."));
+            initBuildStarted = true;
+            return CompletableFuture.completedFuture(null);
+        }
+
+        // stop in-progress compaction tasks to prevent compacted sstable not being indexed.
+        logger.debug(indexContext.logMessage("Stopping active compactions to make sure all sstables are indexed after initial build."));
+        CompactionManager.instance.interruptCompactionFor(Collections.singleton(baseCfs.metadata()),
+                                                          Predicates.alwaysTrue(),
+                                                          true);
+
+        // Force another flush to make sure on disk index is generated for memtable data before marking it queryable.
+        // In case of offline scrub, there is no live memtables.
+        if (!baseCfs.getTracker().getView().liveMemtables.isEmpty())
+        {
+            baseCfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.INDEX_BUILD_STARTED);
+        }
+
+        // It is now safe to flush indexes directly from flushing Memtables.
+        initBuildStarted = true;
+
+        StorageAttachedIndexGroup indexGroup = StorageAttachedIndexGroup.getIndexGroup(baseCfs);
+
+        assert indexGroup != null : "Index group does not exist for table";
```

Comment:
```suggestion
        assert indexGroup != null : "Index group does not exist for table " + baseCfs.keyspace + '.' + baseCfs.name;
```
Aside...how is there no `CFS#getFullyQualifiedName()`?

Reply:
I have wondered this but have tried to hold off adding too much to core classes. I could add one but... :shrug: